clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/loadbalancer/UpsertOpenstackLoadBalancerAtomicOperation.groovy

```diff
@@ -16,203 +16,391 @@
-      method: description.method,
-      subnetId: subnetId,
-      internalPort: description.internalPort)
-    VirtualIP virtualIP = new VirtualIP(
-      name: description.name,
-      subnetId: subnetId,
-      poolId: description.id,
-      protocol: description.protocol,
-      port: description.externalPort)
-    PoolHealthMonitor poolHealthMonitor = description.healthMonitor
-
-    LbPool resultPool
+    LoadBalancerV2 resultLoadBalancer
+
     try {
-      if (newLoadBalancerPool.id) {
-        resultPool = updateLoadBalancer(region, newLoadBalancerPool, virtualIP, poolHealthMonitor)
+      if (!this.description.id) {
+        resultLoadBalancer = createLoadBalancer(region, description.name, description.subnetId)
       } else {
-        resultPool = createLoadBalancer(region, subnetId, newLoadBalancerPool, virtualIP, poolHealthMonitor)
+        resultLoadBalancer = clientProvider.getLoadBalancer(region, description.id)
       }
-    } catch (OpenstackProviderException ope) {
-      throw new OpenstackOperationException(AtomicOperations.UPSERT_LOAD_BALANCER, ope)
-    }
 
-    task.updateStatus UPSERT_LOADBALANCER_PHASE, "Done upserting load balancer ${resultPool?.name} in ${region}"
-    return [(region): [id: resultPool?.id]]
-  }
+      Map<String, ListenerV2> existingListenerMap = buildListenerMap(region, resultLoadBalancer)
+      Map<String, Listener> listenerMap = description.listeners.collectEntries([:]) {
+        [(getListenerKey(it.externalPort, it.externalProtocol.name(), it.internalPort, it.internalProtocol.name())): it]
+      }
+      Map<String, ListenerV2> listenersToUpdate = Maps.newHashMap()
+      Map<String, Listener> listenersToAdd = Maps.newHashMap()
+
+      listenerMap.entrySet()?.each { Map.Entry<String, Listener> entry ->
+        ListenerV2 foundListener = existingListenerMap.get(entry.key)
+        if (foundListener) {
+          listenersToUpdate.put(entry.key, foundListener)
+        } else {
+          listenersToAdd.put(entry.key, entry.value)
+        }
+      }
 
-  protected LbPool createLoadBalancer(String region, String subnetId, LoadBalancerPool newLoadBalancerPool, VirtualIP virtualIP, PoolHealthMonitor poolHealthMonitor) {
-    LbPool resultPool
-    OpenstackClientProvider openstackClientProvider = getClientProvider()
+      Set<String> listenersToDelete = Sets.difference(existingListenerMap.keySet(), Sets.union(listenersToAdd.keySet(), listenersToUpdate.keySet()))
 
-    newLoadBalancerPool.setCreatedTime(System.currentTimeMillis())
+      if (listenersToDelete) {
+        List<ListenerV2> deleteValues = existingListenerMap.findAll { listenersToDelete.contains(it.key) }.collect {
+          it.value
+        }
+        removeListenersAndPools(region, resultLoadBalancer.id, deleteValues)
+      }
 
-    if (!openstackClientProvider.getSubnet(region, subnetId)) {
-      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Unable to retrieve referenced subnet ${subnetId} in ${region}."
```

Comment:
A lot of `task.updateStatus`s have been removed. Do you want to add a couple back where it makes sense?